Sources/Algorithms/Reductions.swift

```diff
@@ -0,0 +1,145 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+/// A sequence of applying a transform to the element of a sequence and the
+/// previously transformed result.
+public struct Reductions<Result, Base: Sequence> {
+  let base: Base
+  let initial: Result
+  let transform: (Result, Base.Element) -> Result
+}
+
+extension Reductions: Sequence {
+  public struct Iterator: IteratorProtocol {
+    var iterator: Base.Iterator
+    var current: Result
+    let transform: (Result, Base.Element) -> Result
+
+    public mutating func next() -> Result? {
+      guard let element = iterator.next() else { return nil }
+      current = transform(current, element)
+      return current
+    }
+  }
+
+  public func makeIterator() -> Iterator {
+    Iterator(iterator: base.makeIterator(),
+             current: initial,
+             transform: transform)
+  }
+}
+
+extension Reductions: Collection where Base: Collection {
+  public struct Index: Comparable {
+    let index: Base.Index
+    let result: Result
+
+    public static func < (lhs: Index, rhs: Index) -> Bool {
+        lhs.index < rhs.index
+    }
+
+    public static func == (lhs: Index, rhs: Index) -> Bool {
+        lhs.index == rhs.index
+    }
+  }
+
+  public var startIndex: Index {
+    let start = base.startIndex
+    let result = transform(initial, base[start])
+    return Index(index: start, result: result)
+  }
+
+  public var endIndex: Index {
+    let end = base.endIndex
+    let result = base.reduce(initial, transform)
+    return Index(index: end, result: result)
+  }
+
+  public subscript(position: Index) -> Result {
+    position.result
+  }
+
+  public func index(after i: Index) -> Index {
+    let index = base.index(after: i.index)
+    let result = transform(i.result, base[i.index])
+    return Index(index: index, result: result)
+  }
+
+  public func distance(from start: Index, to end: Index) -> Int {
+    base.distance(from: start.index, to: end.index)
+  }
+}
+
+extension Reductions: LazyCollectionProtocol
+where Base: LazyCollectionProtocol {}
+
+extension Reductions: LazySequenceProtocol
+  where Base: LazySequenceProtocol {}
+
+extension Collection {
+  public func reductions(
+    _ transform: (Element, Element) throws -> Element
+  ) rethrows -> [Element] {
+
+    guard let first = first else { return [] }
+    return try dropFirst().reductions(first, transform)
+  }
```

Comment:
In order to add this to `Sequence` without having to duplicate any logic, you could take the iterator, call `next` on it to get the seed, and then turn the iterator back into a sequence using [`IteratorSequence`](https://developer.apple.com/documentation/swift/iteratorsequence) (which hopefully optimizes well).